File: 1325-delete-leaves-with-a-given-value/1325-delete-leaves-with-a-given-value.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void helper(TreeNode* root,  int k, bool left, TreeNode* par){
        if(!root) return;
        helper(root->left, k, 1, root);
        helper(root->right, k, 0, root);

        if(!root->left && !root->right && root->val == k){
            if(left) par->left = NULL;
            else par->right = NULL;
        }
    }
    TreeNode* removeLeafNodes(TreeNode* root, int target) {

        helper(root->left, target, 1, root);
        helper(root->right, target, 0, root);
        if(!root->left && !root->right && root->val == target) return NULL;

        return root;
    }
};

File: 1325-delete-leaves-with-a-given-value/1325-delete-leaves-with-a-given-value.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def helper(self, root, k, lft, par):
        if(root == None):
            return 

        self.helper(root.left, k, 1, root)
        self.helper(root.right, k, 0, root)

        if(root.left == None and root.right == None and root.val == k):
            if(lft):
                par.left = None
            else:
                par.right = None

    def removeLeafNodes(self, root: Optional[TreeNode], target: int) -> Optional[TreeNode]:
        self.helper(root.left, target, 1, root)
        self.helper(root.right, target, 0, root)

        if(root.left == None and root.right == None and root.val == target):
            return None
        
        return root